vpr/src/place/placer.cpp

```diff
@@ -0,0 +1,435 @@
+                                                             atom_ctx.nlist,
+                                                             atom_ctx.lookup,
+                                                             *timing_info_->timing_graph(),
+                                                             is_flat_);
+
+   // First time compute timing and costs, compute from scratch
+   PlaceCritParams crit_params;
+   crit_params.crit_exponent = placer_opts_.td_place_exp_first;
+   crit_params.crit_limit = placer_opts_.place_crit_limit;
+
+   initialize_timing_info(crit_params, place_delay_model_.get(), placer_criticalities_.get(),
+                          placer_setup_slacks_.get(), pin_timing_invalidator_.get(),
+                          timing_info_.get(), &costs_, placer_state_);
+
+   critical_path_ = timing_info_->least_slack_critical_path();
+
+   // Write out the initial timing echo file
+   if (isEchoFileEnabled(E_ECHO_INITIAL_PLACEMENT_TIMING_GRAPH)) {
+       tatum::write_echo(getEchoFileName(E_ECHO_INITIAL_PLACEMENT_TIMING_GRAPH),
+                         *timing_ctx.graph, *timing_ctx.constraints,
+                         *placement_delay_calc_, timing_info_->analyzer());
+
+       tatum::NodeId debug_tnode = id_or_pin_name_to_tnode(analysis_opts.echo_dot_timing_graph_node);
+
+       write_setup_timing_graph_dot(getEchoFileName(E_ECHO_INITIAL_PLACEMENT_TIMING_GRAPH) + std::string(".dot"),
+                                    *timing_info_, debug_tnode);
+   }
+
+   costs_.timing_cost_norm = 1 / costs_.timing_cost;
+}
+
+void Placer::check_place_() {
+   const ClusteredNetlist& clb_nlist = g_vpr_ctx.clustering().clb_nlist;
+   const DeviceGrid& device_grid = g_vpr_ctx.device().grid;
+   const auto& cluster_constraints = g_vpr_ctx.floorplanning().cluster_constraints;
+
+   int error = 0;
+
+   // Verify the placement invariants independent to the placement flow.
+   error += verify_placement(placer_state_.blk_loc_registry(),
+                             clb_nlist,
+                             device_grid,
+                             cluster_constraints);
+
+   error += check_placement_costs_();
+
+   if (noc_opts_.noc) {
+       // check the NoC costs during placement if the user is using the NoC supported flow
+       error += noc_cost_handler_->check_noc_placement_costs(costs_, PL_INCREMENTAL_COST_TOLERANCE, noc_opts_);
+       // make sure NoC routing configuration does not create any cycles in CDG
+       error += (int)noc_cost_handler_->noc_routing_has_cycle();
+   }
+
+   if (error == 0) {
+       VTR_LOG("\n");
+       VTR_LOG("Completed placement consistency check successfully.\n");
+
+   } else {
+       VPR_ERROR(VPR_ERROR_PLACE,
+                 "\nCompleted placement consistency check, %d errors found.\n"
+                 "Aborting program.\n",
+                 error);
+   }
+}
+
+int Placer::check_placement_costs_() {
+   int error = 0;
```

Comment:
Small nitpick. I would make this an unsigned! I doubt we can ever have negative errors.